### ISSUE TITLE
Ref identifiers for JsonSubTypes using WRAPPER_OBJECT

### ISSUE THREAD

**Issue by chrisknoll** [victools/jsonschema-generator#165]
Hello, devs,

Firstly, great tool. I needed a way to produce a JSON schema for my object model and this tool worked right away, and invoking the generation directly from a maven task was so nice.

I have a question regarding the generated schema, specifically the keys specified in the `$defs` collection:.

Given this Jackon annotation for specifying the subtypes:

```
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = As.WRAPPER_OBJECT)
@JsonSubTypes({
  @JsonSubTypes.Type(value = ConditionEra.class, name = "ConditionEra"),
  @JsonSubTypes.Type(value = ConditionOccurrence.class, name = "ConditionOccurrence"),
  @JsonSubTypes.Type(value = Death.class, name = "Death"),
  @JsonSubTypes.Type(value = DeviceExposure.class, name = "DeviceExposure"),
  @JsonSubTypes.Type(value = DoseEra.class, name = "DoseEra"),
  @JsonSubTypes.Type(value = DrugEra.class, name = "DrugEra"),
  @JsonSubTypes.Type(value = DrugExposure.class, name = "DrugExposure"),
  @JsonSubTypes.Type(value = LocationRegion.class, name = "LocationRegion"),
  @JsonSubTypes.Type(value = Measurement.class, name = "Measurement"),
  @JsonSubTypes.Type(value = Observation.class, name = "Observation"),
  @JsonSubTypes.Type(value = ObservationPeriod.class, name = "ObservationPeriod"),
  @JsonSubTypes.Type(value = ProcedureOccurrence.class, name = "ProcedureOccurrence"),
  @JsonSubTypes.Type(value = Specimen.class, name = "Specimen"),
  @JsonSubTypes.Type(value = VisitOccurrence.class, name = "VisitOccurrence"),
  @JsonSubTypes.Type(value = PayerPlanPeriod.class, name = "PayerPlanPeriod")
})
public abstract class Criteria {

```
So, there's a base class `Criteria` with a number of subtypes.

Example of where this might be used, is we have a class that defines a proerty that is `Criteria[]`:

```
public class PrimaryCriteria {
  @JsonProperty("CriteriaList")  
  public Criteria[] criteriaList = new Criteria[0];
```

The JSON that would serialize from this field is:
```
CriteriaList: [
  { DrugExposure: {...drug exposure json serialization}},
  { ConditionOccurrence : {...condition occurrence json serialization}}
]
```

The reason I like the WRAPPER_OBJECT mode of serialization is that the serialized json form is exactly what the native type is without any `type` properties co-mingled with it that I think makes it distracting.

My problem:

Looking at the generated json schema, we see this:

```
    "PrimaryCriteria" : {
      "type" : "object",
      "properties" : {
        "CriteriaList" : {
          "type" : "array",
          "items" : {
            "anyOf" : [ {
              "$ref" : "#/$defs/ConditionOccurrence-2"
            , {
              "$ref" : "#/$defs/DrugExposure-2"
            }, { // etc...
```

What is ConditionOccurrence-2?  Is there multiple ConditionOccurrences classes?  Not exactly: the json schema defines a concrepte type to represent the 'wrapper' for the subtypes. This makes complete sense, but I do wish there was either some control over how the wrapper classes were named, or if they just handled 'object wrappers' differently...Like, name `ConditionOccurrence-2` `ConditionOccurrenceWrapper`.  

Alternatively, it would be cool if the `criteria` type was the thing that represented the `anyOf` choices, and the schema produced this:

```
    "Criteria" : {
       "anyOf" : [ {
              "$ref" : "#/$defs/ConditionOccurrence"
            , {
              "$ref" : "#/$defs/DrugExposure"
            }, { // etc...
    "PrimaryCriteria" : {
      "type" : "object",
      "properties" : {
        "CriteriaList" : {
          "type" : "array",
          "items" : {
            "$ref" : "#/$defs/Criteria"
            }
```

Apologies for my hand-waving around the Criteria def:  I wasnt' sure how to articulate that the criteria can have the properties "ConditionOccurrence" or "DrugExposure" and specify the types each property could be.  But I hope the idea is clear.

But, the simples thing for me would be if I could get rid of `ConditionOccurrence-1` and `ConditionOccurrence-2` and instead give them names that closer represent thieir purpose (ie -1 is the physical class so call it `ConditionOccurrence` and -2 is the wrapper).  I feel like it is something to do with [this part of the generator](https://github.com/victools/jsonschema-generator/blob/3fac9c769f772037167c3c9ac8f0f42ba75bcefd/jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JsonSubTypesResolver.java#L224) but I don't think I have the chops to suggest a change.

Also, just would like to add, the other reason I'm trying to fix the names is that in some other cases in the schema, there are fields that directly reference the types directly. I'd like to say the property 'condition' is `$ref` ConditionOcccurrence, and not ConditionOccurrence-1.

**Comment by CarstenWickner** [comment]
Hi @chrisknoll,

I'm glad the tool is useful.
I'll try to respond to your questions one by one:
### 1. Why is there `ConditionOccurrence-1` and `ConditionOccurrence-2` if I only have one class with that name?
That is caused by the way the definitions are constructed internally, but it boils down to what you're already saying yourself: one definition represents the wrapper and the other represents the actual type.
Technically, these are two separate definitions for the same class.
1. One is the (wrapper) definition that will be inserted wherever a property of that type is encountered. The piece of code in the `JacksonModule` you linked to is part of a `CustomDefinitionProviderV2` realising that. Internally, it then makes the following call, which causes a second definition to be created:
```java
context.createStandardDefinitionReference(javaType, this)
```
2. The (inner) definition is being created by the above call, where the `this` parameter refers to the `CustomDefinitionProviderV2` being ignored – otherwise, we'd end up in an endless loop with those wrappers.

Definitions are internally referred to by the class/type they represent. In this special scenario, there are multiple definitions for the same type – the only difference being whether they are the default (wrapper) definition or a nested one, that ignores a custom definition provider.

By default, I kept the logic for avoiding duplicate names simple by just appending an index. For someone who only needs the schema for validation or other purposes where the technical references are irrelevant, this should be good enough.
But I'm aware some people (including you) want to have meaningful definition keys, which brings me to the next point.

### 2. How to change the generated `$defs` keys?
I'm happy to let you know, that there already is a way to configure how each definition's key should be named.
You can find it under the "Individual Configurations" in the documentation:
- https://victools.github.io/jsonschema-generator/#names-in-global-quot-defs-quot-quot-definitions-quot

There are three aspects that you can override, depending on what exactly it is you want to control, but using the `DefaultSchemaDefinitionNamingStrategy` as a basis would be the convenient choice.
```java
configBuilder.forTypesInGeneral().withDefinitionNamingStrategy(new DefaultSchemaDefinitionNamingStrategy() { ... });
```

### 3. How exactly can I give these multiple definitions for one class, meaningful names?
Try this:
```java
configBuilder.forTypesInGeneral().withDefinitionNamingStrategy(new DefaultSchemaDefinitionNamingStrategy() {
    @Override
    public void adjustDuplicateNames(Map<DefinitionKey, String> subschemasWithDuplicateNames, SchemaGenerationContext generationContext) {
        // the definition that does not ignore any custom definition provider is the default/wrapper definition
        List<Map.Entry<DefinitionKey, String>> defaultDefinitions = subschemasWithDuplicateNames.entrySet().stream()
                .filter(entry -> entry.getKey().getIgnoredDefinitionProvider() == null)
                .collect(Collectors.toList());
        if (defaultDefinitions.size() == 1) {
            Map.Entry<DefinitionKey, String> wrapperEntry = defaultDefinitions.get(0);
            wrapperEntry.setValue(wrapperEntry.getValue() + "Wrapper");
            return;
        }
        // if there are multiple default definitions, that means there are multiple classes with the same name in different packages
        super.adjustDuplicateNames(subschemasWithDuplicateNames, generationContext);
    }
});
```
This will result in the `§defs` keys: `ConditionOccurrenceWrapper` and `ConditionOccurrence`.

----

### My question back to you
How do you tell Jackson when to include the subtype _without it being wrapped_?
I believe that's not covered yet on my end, as I always get the wrapper even for a direct reference.

**Comment by chrisknoll** [comment]
Thanks very much for your fast and in-depth reply.  It makes sense but I'd just like to point out that I do understand that the 2 definitions (one for the core class and one for the wrapper) seem to be required in order for the json to work since the form with the wrapper is a sort of 'annonymous class' that is used to identify the type of  subclass that is actually stored in the field.  But, if you were to try to create java classes from the JSON schema, I believe the output classes would create one additinal 'wrapper class' per each instance of a subclass of `criteria` which doesn't need to exist on the java side, and it might look quite confusing to have those.  But that's sorta an artifact of how JSON notation `{ SubClassName: {...SubClassState...}} handles polymorphism....It uses a wrapper with a specific field name in it to determine the class instead of having a flat `type` field in it to do that.  As I eluded to earlier, I like having the wrapper because the json that is being wrapped is the actual serialized form of the underlying object, wheras if i used `type`, the serlized form would have all those fields plus a `type` field.  Which, doesn't actually make sense when polymorphism is at play (which leads us to your question):

> How do you tell Jackson when to include the subtype without it being wrapped?

So the wrapper only comes into play when a super-class reference is specified in the class. Example, consider the following class def (in Java):

```
Criteria[] criteriaList = new Criteria[];
ConditionOccurrence[] coList = new ConditionOccurrenceList[];

```

If I have two ConditionOccurrence instances, and I put them into each of the arrays above, this is the resulting json from Jackson (via the `@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = As.WRAPPER_OBJECT)` annotaiton):

```
{
  criteriaList : [
    {
      ConditionOccurrence : {
          ..... Condition Occurrence 1 State. ...
      }
    }, {
      ConditionOccurrence : {
          ..... Condition Occurrence 2 State. ...
      }
    }
  ]
  coList : [
    {
       ..... Condition Occurrence 1 State. ...
    },{
       ..... Condition Occurrence 2 State. ...
    }
  ]
}
```

Jackson know that since the data type is `Criteria` in the first array, it will apply the wrapper to the elements in the array to provide type information.  IN the case of the second array, three is no polymorphism here (the array is defined as a ConditionOccurrence array, no other type can be in there.

The beauty of the wrapper is the content `...Condition Occurrence State ...` is the same JSON structure within the wrapper as it is as a stand-alone instance, vs, sometimes 'type' would be included (in the case it's a polymorphic context) vs you wouldn't need 'type' because the type is specified in the schema.

So maybe this is just a limitation of JSON Schemas, but ideally, you can define a base type and in your 'definitions' you can say that some subtype is a subclas sof a base type, and you can say things like 'array of BaseType' but any subtype would be valid to be contained in the field.  But, from my perspecitve, I feel the JSON schema should reflect that there is a 'Criteria' type, and there are things like ConditionOccurrences which are a type of 'Criteria', and it is valid to store the ConditionOccurrence structure inside any type that is of `Criteria`.

I think I am also realizing that the notion of this 'wrapper' is a completley Jackson API artifact, and not something explicitly supported by JSONSchema so maybe the reality is that the Jackson generation of schema can only handle what the spec supports...and maybe inheritance isn't specifically supported in the spec and the workaround in Jackson is the wrapped class....

**Comment by CarstenWickner** [comment]
At this stage, I'm not too concerned with any code generation based on the schema. For that, the "Wrapper Object" strategy is simply not an ideal choice. JSON Schema simply is not made for Java's polymorphism as such or any of these oddities. I guess one would have to add some custom properties to tell code generators how to handle such a wrapper definition.

For the fact that the wrapper object or wrapper array is only applied when a subtype is declared, it appears a fix is required.
I'll have to look into that.

**Comment by chrisknoll** [comment]
> For the fact that the wrapper object or wrapper array is only applied when a subtype is declared, it appears a fix is required.
> I'll have to look into that.

I should confirm that behavior before calling it a bug.  But I do like the fact that Jackson applies the wrapper when it's a polymorphic reference (ie: a base type) vs concrete types where no polymorphism is at play is declare directly as the target type. So, I'm actually hoping that isn't something that changes.

**Comment by chrisknoll** [comment]
I greatly apologize, but my information is wrong. I just created a test case that I'll share here, but the TL;DR is that any object that is a `@JsonTypeInfo` subclass _always_ is serialized the same way.  I'm sorry to have the wrong information here: regardless of what the 'reference variable' is, it will always use the wrapped form to render the object into JSON.

Simple class:
```
  public class MultipleTypes {
    public ConditionOccurrence[] coList;
    public Criteria[] criteriaList;
    public ConditionOccurrence co;
    public Criteria criteria;
  }
```

The simple test:
```
    ConditionOccurrence c1 = new ConditionOccurrence();
    c1.codesetId = 1;
    ConditionOccurrence c2 = new ConditionOccurrence();
    c2.codesetId = 2;
    
    MultipleTypes mt = new MultipleTypes();
    mt.co = c1;
    mt.coList = new ConditionOccurrence[] {c1, c2};
    mt.criteria = c1;
    mt.criteriaList = new Criteria[] {c1, c2};
    
    String json = Utils.serializePretty(mt);
```

The formatted json output:

```
{
  "coList": [
    {
      "ConditionOccurrence": {
        "CodesetId": 1,
        "ConditionTypeExclude": false
      }
    },
    {
      "ConditionOccurrence": {
        "CodesetId": 2,
        "ConditionTypeExclude": false
      }
    }
  ],
  "criteriaList": [
    {
      "ConditionOccurrence": {
        "CodesetId": 1,
        "ConditionTypeExclude": false
      }
    },
    {
      "ConditionOccurrence": {
        "CodesetId": 2,
        "ConditionTypeExclude": false
      }
    }
  ],
  "co": {
    "ConditionOccurrence": {
      "CodesetId": 1,
      "ConditionTypeExclude": false
    }
  },
  "criteria": {
    "ConditionOccurrence": {
      "CodesetId": 1,
      "ConditionTypeExclude": false
    }
  }
}
```

So, regardless of what the variable reference type is, the wrapper is always applied.  Makes sense.  But this wrapper is not part of the underlying object model, just a construct to handle the polymorphism in JSON.  This is not really avoidable (when using the wrapper form) because the type information needs to be represented somewhere. That also makes sense.

I think @CarstenWickner code example is worth looking into because even tho these wrapper classes need to be represented somehow in the JSON schema, I think it gives me the option to try to 'project' them as a different name than the ClassName-2.  

Thank you all so much for your time and attention, it was very helpful, and I do sincerely apologize for my wrong information.

**Comment by chrisknoll** [comment]
Follow up question: now that we established that subtypes of classes marked with `@JsonTypeInfo` are never serialized without a wrapper, I wonder if there may be an improvement that would place the definition of the subtime as a single declaration (because you can't ever format any of the subtypes as non wrapped):

Current schema def:

```
    "ConditionOccurrence-1" : {
      "type" : "object",
      "properties" : {
        "Age" : {
          "$ref" : "#/definitions/NumericRange"
        },
        "CodesetId" : {
          "type" : "integer"
        }
      }
    },
    "ConditionOccurrence-2" : {
      "type" : "object",
      "properties" : {
        "ConditionOccurrence" : {
          "$ref" : "#/definitions/ConditionOccurrence-1"
        }
      }
    }
```

Possible alternative (for clarity:
```
    "ConditionOccurrence" : {
      "type" : "object",
      "properties" : {
        "ConditionOccurrence" : {
          "type" : "object",
          "properties" : {
            "Age" : {
              "$ref" : "#/definitions/NumericRange"
            },
            "CodesetId" : {
              "type" : "integer"
            }
          }
        }
      }
```

So the definition of the ConditionOccuence implies the wrapper (which it does due to the Jackson annotation) and, to me, it's more clear.  Still makes it hard to create classes from the schema, but I agree with you, that's a secondary concern. A tool that is built to create java classes from this would have to recognize the wrapped pattern and identify this as something with `JsonTypeInfo`, but no idea how it would figure out a base class from that.  But, that's a problem for another repo!

**Comment by CarstenWickner** [comment]
Hi @chrisknoll,

You were too quick to apologise. Jackson does support a per-property override of the `@JsonTypeInfo`, e.g. to `@JsonTypeInfo(use = JsonTypeInfo.Id.NONE)`

#### Example:
```java
    static class Parent {
        public AnimalData animalData;
        @JsonTypeInfo(use = JsonTypeInfo.Id.NONE)
        public Human directHuman;
    }

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.WRAPPER_OBJECT)
    @JsonSubTypes({ @JsonSubTypes.Type(value = Human.class, name = "human"),
            @JsonSubTypes.Type(value = Frog.class, name = "frog"), })
    static class AnimalData{
        public long foo;
    }

    @JsonTypeName("human")
    static class Human extends AnimalData {
        public String bar;
    }

    @JsonTypeName("frog")
    static class Frog extends AnimalData {
        public int baz;
    }
```
#### Test Code
```java
    Parent example = new Parent();
    example.directHuman = new Human();
    example.directHuman.foo = 7800000000L;
    example.directHuman.bar = "2000-01-01";
    example.animalData = example.directHuman;
    System.out.println(config.getObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(example));
```
#### Test Output
```json
{
  "animalData" : {
    "human" : {
      "foo" : 7800000000,
      "bar" : "2000-01-01"
    }
  },
  "directHuman" : {
    "foo" : 7800000000,
    "bar" : "2000-01-01"
  }
}
```

The above is currently not supported in the `jsonschema-generator`.

**Comment by chrisknoll** [comment]
I see, thanks for checking. Personally, i wouldn't toggle that option to use `Id.NONE` because if they are always wrapped, you can represent the object consistently across all contexts, while using it means that you'd have to actually define 2 different schemas (one wrapped, one not).  So...very cool and interesting flexibility offered there, but, man, makes your life harder to support all those options.

Again, very much appreciate your feedback and comments, and I hope to enjoy using the library.

**Comment by CarstenWickner** [comment]
I've kept this issue open for a little over a year now, because of the lacking support regarding the `@JsonTypeInfo(use = JsonTypeInfo.Id.NONE)` property override option.

I'm closing this now, as I finally got around to making that (admittedly: small) change. It'll be part of the next release (4.24.0) then – not clear yet when that'll be released as long as nobody asks for it and it's the only change for that version so far.
